### PR TITLE
mem-ruby,tests: Add CHI with ISA tests

### PR DESCRIPTION
**Pull request by BobbyRBruce**
These tests ensure CHI is sitll working with ARM, X86, and RISCV ISA targets. Additionally whether it still functions in atomic mode and with varying number of cores.

### CODE REVIEW

**Comment by BobbyRBruce** [comment]
@hnpl : You requested tests for a test running X86 with CHI. This is the solution I came up with. However, all these tests are failing with:

```txt
Assertion failed: (x > 0), function floorLog2, file intmath.hh, line 61.
Program aborted at tick 333
--- BEGIN LIBC BACKTRACE ---
0   gem5.opt                            0x0000000104fd1dd0 _ZN4gem515print_backtraceEv + 44
1   gem5.opt                            0x0000000104fe73c8 _ZN4gem512abortHandlerEi + 132
2   libsystem_platform.dylib            0x0000000193ff0184 _sigtramp + 56
3   libsystem_pthread.dylib             0x0000000193fbaf70 pthread_kill + 288
4   libsystem_c.dylib                   0x0000000193ec7908 abort + 128
5   libsystem_c.dylib                   0x0000000193ec6c1c err + 0
6   gem5.opt                            0x0000000105613118 _ZN4gem54ruby16Cache_Controller16processNextStateERKyPNS0_9Cache_TBEEPNS0_16Cache_CacheEntryE.cold.1 + 0
7   gem5.opt                            0x000000010542f278 _ZN4gem54ruby16Cache_Controller21Deallocate_CacheBlockERPNS0_9Cache_TBEERPNS0_16Cache_CacheEntryEy + 0
8   gem5.opt                            0x00000001054004d8 _ZN4gem54ruby16Cache_Controller11getDirEntryERKy + 40
9   gem5.opt                            0x00000001054001a4 _ZN4gem54ruby16Cache_Controller8getStateEPNS0_9Cache_TBEEPNS0_16Cache_CacheEntryERKy + 48
10  gem5.opt                            0x0000000105461554 _ZN4gem54ruby16Cache_Controller12doTransitionENS0_11Cache_EventEPNS0_16Cache_CacheEntryEPNS0_9Cache_TBEEy + 60
11  gem5.opt                            0x000000010546a434 _ZN4gem54ruby16Cache_Controller6wakeupEv + 6136
12  gem5.opt                            0x00000001054e9a18 _ZN4gem54ruby8Consumer19processCurrentEventEv + 256
13  gem5.opt                            0x0000000104fde4f8 _ZN4gem510EventQueue10serviceOneEv + 244
14  gem5.opt                            0x0000000104ff8e8c _ZN4gem59doSimLoopEPNS_10EventQueueE + 272
15  gem5.opt                            0x0000000104ff8898 _ZN4gem58simulateEy + 408
16  gem5.opt                            0x0000000102bb510c _ZZN8pybind1112cpp_function10initializeIRPFPN4gem522GlobalSimLoopExitEventEyES4_JyEJNS_4nameENS_5scopeENS_7siblingENS_5arg_vEEEEvOT_PFT0_DpT1_EDpRKT2_ENKUlRNS_6detail13function_callEE_clESP_ + 84
17  gem5.opt                            0x0000000102bb50ac _ZZN8pybind1112cpp_function10initializeIRPFPN4gem522GlobalSimLoopExitEventEyES4_JyEJNS_4nameENS_5scopeENS_7siblingENS_5arg_vEEEEvOT_PFT0_DpT1_EDpRKT2_ENUlRNS_6detail13function_callEE_8__invokeESP_ + 24
18  gem5.opt                            0x0000000102b57f0c _ZN8pybind1112cpp_function10dispatcherEP7_objectS2_S2_ + 5212
19  Python                              0x0000000109e05430 cfunction_call + 76
20  Python                              0x0000000109da8a44 _PyObject_Call + 128
21  Python                              0x0000000109ed05b4 _PyEval_EvalFrameDefault + 13024
22  Python                              0x0000000109ecd07c PyEval_EvalCode + 200
23  Python                              0x0000000109ec8688 builtin_exec + 444
24  Python                              0x0000000109e04af4 cfunction_vectorcall_FASTCALL_KEYWORDS + 88
25  Python                              0x0000000109da88c8 PyObject_Vectorcall + 92
26  Python                              0x0000000109ecf610 _PyEval_EvalFrameDefault + 9020
27  gem5.opt                            0x0000000102ba305c _ZNK8pybind116detail10object_apiINS0_8accessorINS0_17accessor_policies8str_attrEEEEclILNS_19return_value_policyE1EJEEENS_6objectEDpOT0_ + 76
28  gem5.opt                            0x0000000104fe7eb8 main + 344
29  dyld                                0x0000000193c38274 start + 2840
--- END LIBC BACKTRACE ---
```

Am I doing something stupid with my setup here or is this a bug?

**Comment by hnpl** [comment]
Your setup looks fine. The error is somewhere in one of the cache structs,

```
#8  0x000055555bfeff5a in gem5::ruby::PerfectCacheMemory<gem5::ruby::Cache_DirEntry>::isTagPresent (this=0x5555619f3900, address=7168)
    at /scr/hn/gem5/src/mem/ruby/structures/PerfectCacheMemory.hh:137
137         return m_map.count(makeLineAddress(address, floorLog2(m_block_size))) > 0;
(gdb) print m_block_size
$1 = 0
```

**Comment by powerjg** [comment]
Probably related to #1453. Tagging @abmerop 

My guess is that you need to pass the block size to the directory memory object.

**Comment by abmerop** [comment]
> Probably related to #1453. Tagging @abmerop
> 
> My guess is that you need to pass the block size to the directory memory object.

What build is this? I could try testing as well. My guess is PerfectCacheMemory needs to be added to the list of objects that needs setBlockSize to be called (https://github.com/gem5/gem5/blob/develop/src/mem/slicc/symbols/StateMachine.py#L776)

**Comment by hnpl** [comment]
This is the build with the CHI protocol. This is the build_opts file,

```
RUBY=y
RUBY_PROTOCOL_CHI=y
BUILD_ISA=y
USE_ARM_ISA=y
USE_RISCV_ISA=y
USE_X86_ISA=y
USE_TEST_OBJECTS=y
```

**Comment by BobbyRBruce** [comment]
FYI: I have no solution to this. I'll likely create an issue soon but I don't have the bandwidth (or Ruby/CHI expertise to fix this). Anyone is free to jump in and try to get these tests passing.

**Comment by BobbyRBruce** [comment]
to reproduce (with this PR checked out):

```shell
cd tests
./main.py run -j `nproc` --length=long -vvv gem5/chi_protocol
```

(This will build the `ALL_CHI` binary for you. With `-vvv` you'll get the command run for each test as part of the output).

**Comment by BobbyRBruce** [comment]
This same bug is causing the Daily ARM boot tests to fail (when using Chi): https://github.com/gem5/gem5/actions/runs/11391775359/job/31696319075

**Comment by abmerop** [comment]
> This same bug is causing the Daily ARM boot tests to fail (when using Chi): https://github.com/gem5/gem5/actions/runs/11391775359/job/31696319075

Thanks will take a look this weekend

**Comment by hnpl** [comment]
> > Probably related to #1453. Tagging @abmerop
> > My guess is that you need to pass the block size to the directory memory object.
> 
> What build is this? I could try testing as well. My guess is PerfectCacheMemory needs to be added to the list of objects that needs setBlockSize to be called (https://github.com/gem5/gem5/blob/develop/src/mem/slicc/symbols/StateMachine.py#L776)

Yeah, I think you're right. I looked into the slicc code, and I think this is the right place to add the block size to the PerfectCacheMemory object. Though, a different error arose after patching this error.

```
gem5.debug: src/mem/ruby/common/NetDest.cc:281: bool gem5::ruby::NetDest::isElement(gem5::ruby::MachineID) const: Assertion `m_bits.size() > 0' failed.
Program aborted at tick 333
--- BEGIN LIBC BACKTRACE ---                                                                                                                                                                                                                
build/ARM/gem5.debug(_ZN4gem515print_backtraceEv+0x32)[0x63a6f4eda0cf]                                                                                                                                                                      
build/ARM/gem5.debug(_ZN4gem512abortHandlerEi+0x7e)[0x63a6f4f0019e]  
/lib/x86_64-linux-gnu/libc.so.6(+0x45320)[0x7d150e645320]
/lib/x86_64-linux-gnu/libc.so.6(pthread_kill+0x11c)[0x7d150e69eb1c]                                                                                                                                                                         
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0x1e)[0x7d150e64526e]
/lib/x86_64-linux-gnu/libc.so.6(abort+0xdf)[0x7d150e6288ff]                  
/lib/x86_64-linux-gnu/libc.so.6(+0x2881b)[0x7d150e62881b]                                                                                                                                                                                   
/lib/x86_64-linux-gnu/libc.so.6(+0x3b507)[0x7d150e63b507]                                                             
build/ARM/gem5.debug(_ZNK4gem54ruby7NetDest9isElementENS0_9MachineIDE+0x4e)[0x63a6f46d4190]                                                                                                                                                 
build/ARM/gem5.debug(_ZN4gem54ruby16Cache_Controller23_checkResourceAvailableERKNS0_17Cache_RequestTypeERKm+0x25b)[0x63a6f4435e9d]
build/ARM/gem5.debug(_ZN4gem54ruby16Cache_Controller22checkResourceAvailableERKNS0_17Cache_RequestTypeERKm+0x48)[0x63a6f443600e]                                                                                                            
build/ARM/gem5.debug(_ZN4gem54ruby16Cache_Controller18doTransitionWorkerENS0_11Cache_EventENS0_11Cache_StateERS3_RPNS0_9Cache_TBEERPNS0_16Cache_CacheEntryEm+0x540d)[0x63a6f449324b]
build/ARM/gem5.debug(_ZN4gem54ruby16Cache_Controller12doTransitionENS0_11Cache_EventEPNS0_16Cache_CacheEntryEPNS0_9Cache_TBEEm+0x217)[0x63a6f448d139]
build/ARM/gem5.debug(_ZN4gem54ruby16Cache_Controller6wakeupEv+0x1fb5)[0x63a6f449bc63]
build/ARM/gem5.debug(_ZN4gem54ruby8Consumer19processCurrentEventEv+0xd3)[0x63a6f46cec85]
build/ARM/gem5.debug(+0x5aa5707)[0x63a6f46ce707]                                                                                                                                                                                            
build/ARM/gem5.debug(+0x5aa6006)[0x63a6f46cf006]                                                                      
build/ARM/gem5.debug(+0x5aa5eed)[0x63a6f46ceeed]                                                                                                                                                                                            
build/ARM/gem5.debug(+0x5aa5ddb)[0x63a6f46ceddb]                                                                      
build/ARM/gem5.debug(_ZNKSt8functionIFvvEEclEv+0x36)[0x63a6f27163fe]
build/ARM/gem5.debug(_ZN4gem520EventFunctionWrapper7processEv+0x20)[0x63a6f279fa0a]
build/ARM/gem5.debug(_ZN4gem510EventQueue10serviceOneEv+0x121)[0x63a6f4eedc3d]
build/ARM/gem5.debug(_ZN4gem59doSimLoopEPNS_10EventQueueE+0x23d)[0x63a6f4f1486b]
build/ARM/gem5.debug(_ZN4gem58simulateEm+0x25e)[0x63a6f4f142d0]
build/ARM/gem5.debug(+0x3aaa5a8)[0x63a6f26d35a8]
build/ARM/gem5.debug(+0x3aa8432)[0x63a6f26d1432]
build/ARM/gem5.debug(+0x3aa3fe3)[0x63a6f26ccfe3]
build/ARM/gem5.debug(+0x3aa404e)[0x63a6f26cd04e]
build/ARM/gem5.debug(+0x3996f2f)[0x63a6f25bff2f]
/lib/x86_64-linux-gnu/libpython3.12.so.1.0(+0x1dbe88)[0x7d15101dbe88]
/lib/x86_64-linux-gnu/libpython3.12.so.1.0(_PyObject_Call+0x5c)[0x7d151018482c]
/lib/x86_64-linux-gnu/libpython3.12.so.1.0(_PyEval_EvalFrameDefault+0x8ec)[0x7d1510119dec]
--- END LIBC BACKTRACE ---
```

**Comment by hnpl** [comment]
The error is due to the NetDest constructor not being initialized with the RubySystem, so the m_bits vector is empty.

In `build/ARM/mem/ruby/protocol/Cache_Controller.cc`,

```C++
Cache_Controller::init()
{
    ....
    m_destsWaitingRetry_ptr  = new NetDest();
    assert(m_destsWaitingRetry_ptr != NULL);
    ...
}
```

**Comment by BobbyRBruce** [comment]
> The error is due to the NetDest constructor not being initialized with the RubySystem, so the m_bits vector is empty.
> 
> In `build/ARM/mem/ruby/protocol/Cache_Controller.cc`,
> 
> ```c++
> Cache_Controller::init()
> {
>     ....
>     m_destsWaitingRetry_ptr  = new NetDest();
>     assert(m_destsWaitingRetry_ptr != NULL);
>     ...
> }
> ```

Can you push a fix?

**Comment by hnpl** [comment]
yeah, I haven't fixed all the problems though ...

**Comment by abmerop** [comment]
Can you check if #1690 works?  I don't know how long this test should take

**Comment by hnpl** [comment]
@BobbyRBruce @abmerop I pushed a fix #1689. The fix is in progress though, not all problems were addressed.

**Comment by hnpl** [comment]
@abmerop Haha we got a race condition here. I'll check your patch.

**Comment by hnpl** [comment]
I got this error,

```
gem5.debug: src/mem/ruby/common/WriteMask.hh:166: bool gem5::ruby::WriteMask::isEmpty() const: Assertion `mSize > 0' failed.
Program aborted at tick 8914743
--- BEGIN LIBC BACKTRACE ---
build/ARM/gem5.debug(_ZN4gem515print_backtraceEv+0x32)[0x6479e1d409e7]
build/ARM/gem5.debug(_ZN4gem512abortHandlerEi+0x7e)[0x6479e1d66ab6]
/lib/x86_64-linux-gnu/libc.so.6(+0x45320)[0x7ef909445320]
/lib/x86_64-linux-gnu/libc.so.6(pthread_kill+0x11c)[0x7ef90949eb1c]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0x1e)[0x7ef90944526e]
/lib/x86_64-linux-gnu/libc.so.6(abort+0xdf)[0x7ef9094288ff]
/lib/x86_64-linux-gnu/libc.so.6(+0x2881b)[0x7ef90942881b]
/lib/x86_64-linux-gnu/libc.so.6(+0x3b507)[0x7ef90943b507]
build/ARM/gem5.debug(_ZNK4gem54ruby9WriteMask7isEmptyEv+0x42)[0x6479e120fa4e]
build/ARM/gem5.debug(_ZN4gem54ruby17Memory_Controller14functionalReadERKmPNS_6PacketERNS0_9WriteMaskE+0x175)[0x6479e1315d7d]
build/ARM/gem5.debug(_ZN4gem54ruby10RubySystem14functionalReadEPNS_6PacketE+0x64e)[0x6479e13ebcfa]
build/ARM/gem5.debug(_ZN4gem54ruby8RubyPort15MemResponsePort14recvFunctionalEPNS_6PacketE+0x38b)[0x6479e13dfc6b]
build/ARM/gem5.debug(_ZNK4gem525FunctionalRequestProtocol4sendEPNS_26FunctionalResponseProtocolEPNS_6PacketE+0x6a)[0x6479e120cc24]
build/ARM/gem5.debug(_ZNK4gem511RequestPort14sendFunctionalEPNS_6PacketE+0x5f)[0x6479df62ab9f]
build/ARM/gem5.debug(_ZN4gem513ThreadContext14sendFunctionalEPNS_6PacketE+0xa7)[0x6479e18c0417]
build/ARM/gem5.debug(+0x570e280)[0x6479e119d280]
build/ARM/gem5.debug(+0x570f0e2)[0x6479e119e0e2]
build/ARM/gem5.debug(+0x570ee3c)[0x6479e119de3c]
build/ARM/gem5.debug(+0x570ec00)[0x6479e119dc00]
build/ARM/gem5.debug(_ZNKSt8functionIFvPN4gem56PacketEEEclES2_+0x4d)[0x6479e119e4bd]
build/ARM/gem5.debug(_ZNK4gem59PortProxy12readBlobPhysEmNS_5FlagsImEEPvm+0x13b)[0x6479e119d555]
build/ARM/gem5.debug(+0x575eb0c)[0x6479e11edb0c]
build/ARM/gem5.debug(+0x575f959)[0x6479e11ee959]
build/ARM/gem5.debug(+0x575f539)[0x6479e11ee539]
build/ARM/gem5.debug(+0x575f1e6)[0x6479e11ee1e6]
build/ARM/gem5.debug(_ZNKSt8functionIFvRKN4gem514TranslationGen5RangeEEEclES4_+0x4d)[0x6479e11ef40f]
build/ARM/gem5.debug(_ZNK4gem520TranslatingPortProxy9tryOnBlobENS_7BaseMMU4ModeESt10unique_ptrINS_14TranslationGenESt14default_deleteIS4_EESt8functionIFvRKNS4_5RangeEEE+0x13e)[0x6479e11ed9f4]
build/ARM/gem5.debug(_ZNK4gem520TranslatingPortProxy11tryReadBlobEmPvm+0xca)[0x6479e11edbfa]
build/ARM/gem5.debug(_ZNK4gem59PortProxy8readBlobEmPvm+0x4f)[0x6479df741869]
build/ARM/gem5.debug(_ZN4gem514ProxyPtrBufferINS_22SETranslatingPortProxyEE4loadEv+0xea)[0x6479dfa18604]
build/ARM/gem5.debug(_ZN4gem514ProxyPtrBufferINS_22SETranslatingPortProxyEEC1ESt10shared_ptrIS1_Emm+0x7b)[0x6479dfa17e79]
build/ARM/gem5.debug(_ZN4gem513ConstProxyPtrINS_5Linux7utsnameENS_22SETranslatingPortProxyEE7setAddrEm+0x75)[0x6479dfa16727]
```

**Comment by abmerop** [comment]
`WriteMask` size must be explicitly set now.  I updated my change setting the size in CHI's `functionalRead` functions

**Comment by hnpl** [comment]
I think this need to be fixed as well,
https://github.com/gem5/gem5/blob/b1a44b89c7bae73fae2dc547bc1f871452075b85/src/mem/ruby/system/RubySystem.cc#L673

**Comment by hnpl** [comment]
Do we need to test the atomic mode given that the atomic mode does not use the ruby cache?

File: tests/gem5/chi_protocol/test_chi_per_isa.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+if config.bin_path:
+    resource_directory = config.bin_path
+else:
+    resource_directory = joinpath(
+        config.base_dir, "tests", "gem5", "resources"
+    )
+
+for isa in ("arm", "riscv", "x86"):
+    for num_cores in (1, 2, 4):
+        for atomic in (True, False):
```
Comment (powerjg) [inline tests/gem5/chi_protocol/test_chi_per_isa.py]: We should only test timing. As Hoa said, Ruby doesn't work with atomic.

**Review by powerjg — CHANGES_REQUESTED** [review]

File: tests/gem5/chi_protocol/test_chi_per_isa.py (hunk 1 above)
Comment (BobbyRBruce) [reply tests/gem5/chi_protocol/test_chi_per_isa.py]: I just wanted a test to check it wasn't catastrophically incompatible. I.e., things still run, the simulation didn't crash. Though i'm aware ruby doesn't work with atomic you can still put ruby caches in simulations that are run with atomic cores and doing so shouldn't cause issue. These tests are very fast, so the cost covering the cross-product of possible setups isn't costly. Regardless, the tests may send the wrong message to users (I suppose someone may thing "if they are testing it, it must be a valid config, right?") so I'll remove it.

(tl;dr: ok)

**Review by powerjg — APPROVED** [review]

**Review by hnpl — APPROVED** [review]

**Pull request merged**